### ISSUE TITLE
Add Path.RemoveRelativeSegments Api

### ISSUE THREAD

**Issue by Anipik** [dotnet/runtime#2162]
# Rationale
Currently there is no direct way to normalize the path or to remove the relative Segments from any path string. We can use ```GetFullPath``` to normalize the path but it will resolve it relative to root directory or working directory which is not always desired.
eg

- working with relative paths in archives where we may require normalized paths that are "full" paths in said archive, but not on the system

- Will help in building and working with relative urls.

# Proposed Api
``` C#
namespace System.IO
{
    public class Path
    {
        public static string RemoveRelativeSegments(string path);
        public static string RemoveRelativeSegments(ReadOnlySpan<char> path);
        public static bool TryRemoveRelativeSegments(ReadOnlySpan<char> path, span<char> buffer, out int charsWritten);
    }
}
```
The ```rootlength``` is the length of the path which will never be trimmed while removing the relativeSegments. 

# Behavior

- `Path.DirectorySeparatorChar` and `Path.AltDirectorySeparatorChar` are considered path "segment" separators (`\` and `/` on Windows, `/` on Unix)
- Sequential separators will be collapsed to a single separator
- Any `Path.AltDirectorySeparatorChar` characters will be changed to `Path.DirectorySeparatorChar` (only relevant on Windows).
- Segments of one period only `.` will be removed.
- Segments of two periods only `..` will be removed along with the parent segment, if any.

# Implementation

This api is already implemented as an internal api and is being used by ```GetFullPath``` api to remove relative segments in case of device paths. 
https://github.com/dotnet/runtime/blob/96d9a47228b0412945d151a79af7bf66463b053a/src/libraries/System.Private.CoreLib/src/System/IO/PathInternal.cs#L129

# Usage

```cs
// removing the relative segment
Assert.Equal("C:\temp", PathInternal.RemoveRelativeSegments("C:\git\..\temp"));

// removing the relative segment but not eating the root
Assert.Equal("C:\temp", PathInternal.RemoveRelativeSegments("C:\..\..\temp"));

// Input is relative path
Assert.Equal("temp", PathInternal.RemoveRelativeSegments("git\..\temp"));

// Multiple types of relative Segments together
Assert.Equal("temp", PathInternal.RemoveRelativeSegments("git\.\..\temp"));
Assert.Equal("git\temp", PathInternal.RemoveRelativeSegments("git\\\\temp"));

// Normalizing the path
Assert.Equal("git\temp\src\corefx", PathInternal.RemoveRelativeSegments("git\temp/src\corefx"));
```

More Usages can be find here 
https://github.com/dotnet/corefx/pull/37225
I have modified the internal tests to api proposal

# Some Design Answers

>  Resolving this https://github.com/dotnet/corefx/issues/4208

Can directly call RemoveRelativeSegments(RelativePath);

# Exception Handling 

-   return empty string if the path is empty or null.
-  throw Argument Exception if the bufferLength is less than the required in TryRemoveRelativeSegments api

cc @danmosemsft @JeremyKuhne 

related implementation prs https://github.com/dotnet/coreclr/pull/24273 https://github.com/dotnet/corefx/pull/37225

**Comment by jkotas** [comment]
> public static string RemoveRedundantSegments(ReadOnlySpan<char> path);

Should this return Span, like existing `Path.TrimEndingDirectorySeparator`? Returning string means that this API will allocate unnecessarily when the path does not contain any redundant segments.

**Comment by carlossanlop** [comment]
@jkotas I see what you mean. I am working on this right now and as I was implementing that method, I started wondering about that. You were faster than me in asking about it. 😸 

I did a quick replay of the video, but I couldn't find any mention of the return value, so I suppose it was a small oversight (I could've also missed it).

Question to the reviewers in the video: Should we update the second signature to return a span instead of a string? What's the process to alter an already approved API?

 @terrajobst @bartonjs @GrabYourPitchforks @tannergooding @KrzysztofCwalina @scalablecory @JeremyKuhne

**Comment by bartonjs** [comment]
> Should we update the second signature to return a span instead of a string? 

No, the guidance is to only return a span that is a slice (including the trivially-all slice or trivially-none slice) of an input parameter.

Someone wanting the non-allocating one will just have to use the Try method.  (And/or make is one return null if the input is unchanged, but that might make it harder to use.  The one that accepts a string and returns a string can return the input argument on no redundant segment)

**Comment by jeffhandley** [comment]
I'm late to the game here, but I'm struggling with the word "redundant" in the API name, as I don't think we're matching the [definition of "redundant"](https://www.bing.com/search?q=define+redundant). The path `/foo/bar/../baz` does not have any _redundant_ segments as nothing is repeated or excessive--instead `..` _negates_ `bar`; it cancels it out.

Node has a [`Path.normalize` function](https://nodejs.org/dist/latest-v13.x/docs/api/path.html#path_path_normalize_path) that is strikingly similar to what is being proposed here. I always liked the "normalize" verb for this function. To make this specific to relative path segments, I recommend `NormalizeRelativeSegments`.

**Comment by jeffhandley** [comment]
Moving this out to .NET 7.0.

**Comment by deeprobin** [comment]
Why do we have two overloads for this?
```cs
        public static string RemoveRedundantSegments(string path);
        public static string RemoveRedundantSegments(ReadOnlySpan<char> path);
```

The string could be automatically converted to a `ReadOnlySpan<char>` by the `implicit operator`, or are there any performance advantages expected by using a heap reference type?

---

I would like to take over the issue. So you are welcome to assign me :).

**Comment by deeprobin** [comment]
@carlossanlop What is the current progress for #37939 (closed for inactivity 😩)

**Comment by carlossanlop** [comment]
@deeprobin The closed PR has the feature fully working and properly tested. The only problem is that it would be introduced a performance regression in Unix, which can be reproduced by calling `Path.GetFullPath("/some/existing/path/without/relative/segments")` and comparing the measured time it takes to execute that API, before and after the change. But once we get that regression investigated and fixed, it can get merged.

Right now I don't have bandwidth to work on that, but we would welcome external contributions that can take on the above task and help drive that PR to completion.

Would you be interested, @deeprobin ?

**Comment by deeprobin** [comment]
@carlossanlop I would take a look at it. However, I am currently not sure if I can improve the regression solution.

**Comment by iSazonov** [comment]
Looking #37939 I guess the code could be more simple if handle the path from end to begin.

**Comment by carlossanlop** [comment]
> Looking #37939 I guess the code could be more simple if handle the path from end to begin.

Feel free to copy my branch and submit a PR, @iSazonov .

**Comment by krwq** [comment]
@jeffhandley I've unassigned you given there has been no activity in .NET 7-8.

**Comment by pCYSl5EDgo** [comment]
Hi everyone.

Before preparing a formal PR, I have developed [a proof-of-concept implementation](https://github.com/pCYSl5EDgo/Pcysl5edgo.RemoveRedundantPath) of this api that substantially improves performance while also addressing a small set of normalization inconsistencies in the Windows behavior of #37939 .

Performance summary:
- ~1.25x faster overall compared to #37939
- 6x–10x faster on the specific Unix path (`/some/existing/path/without/relative/segments`, already normalized long full path) that previously triggered regressions.
- Fully matches existing behavior, except for several Windows-only cases where the current implementation produces non-minimal normalized paths (e.g. ".\.." → ".\.." instead of the minimal ".."). 
- All Windows/Unix edge cases have been exhaustively re-tested. (Thank you for the good tests.)

Repository (PoC):  [https://github.com/pCYSl5EDgo/Pcysl5edgo.RemoveRedundantPath](https://github.com/pCYSl5EDgo/Pcysl5edgo.RemoveRedundantPath).

The PoC repository includes:
• The full implementation in a standalone form  
• A complete BenchmarkDotNet suite (old vs new vs `Path.GetFullPath`)  
• Cross-platform tests including corner cases and the scenarios that caused earlier regressions  

If you have concerns about compatibility, root semantics, device path prefixes, or separator collapsing rules, I would be glad to incorporate the feedback before opening the PR.

Thanks.

cc @carlossanlop
